python_modules/libraries/dagster-databricks/README.md

```diff
@@ -2,3 +2,146 @@
+exits gracefully on success or failure:
+
+
+```
+import os
+from dagster import AssetExecutionContext, Definitions, asset
+from dagster_databricks import ExtDatabricks
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.service import jobs
+
+@asset
+def databricks_asset(context: AssetExecutionContext, ext: ExtDatabricks):
+
+    # task specification will be passed to databricks as-is, except for the
+    # injection of environment variables
+    task = jobs.SubmitTask.from_dict({
+        "new_cluster": { ... },
+        "libraries": [
+            # must include dagster-ext-process
+            {"pypi": {"package": "dagster-ext-process"}},
+        ],
+        "task_key": "some-key",
+        "spark_python_task": {
+            "python_file": "dbfs:/myscript.py",
+            "source": jobs.Source.WORKSPACE,
+        }
+    })
+
+    # arbitrary json-serializable data you want access to from the ExtContext
+    # in the databricks runtime
+    extras = {"foo": "bar"}
+
+    # synchronously execute the databricks job
+    ext.run(
+        task=task,
+        context=context,
+        extras=extras,
+    )
+
+client = WorkspaceClient(
+    host=os.environ["DATABRICKS_HOST"],
+    token=os.environ["DATABRICKS_TOKEN"],
+)
+
+defs = Definitions(
+    assets=[databricks_asset],
+    resources = {"ext": ExtDatabricks(client)}
+)
+```
+
+`ExtDatabricks.run` requires that the targeted python script
+(`dbfs:/myscript.py` above) already exist in DBFS. Here is what it might look
+like:
+
+```
+### dbfs:/myscript.py
+
+# `dagster_ext` must be available in the databricks python environment
+from dagster_ext import ExtDbfsContextLoader, ExtDbfsMessageWriter, init_dagster_ext
+
+# Sets up communication channels and downloads the context data sent from Dagster.
+# Note that while other `context_loader` and `message_writer` settings are
+# possible, it is recommended to use the below settings for Databricks.
+context = init_dagster_ext(
+    context_loader=ExtDbfsContextLoader(),
+    message_writer=ExtDbfsMessageWriter()
+)
+
+# Access the `extras` dict passed when launching the job from Dagster.
+foo_extra = context.get_extra("foo")
+
+# Stream log message back to Dagster
+context.log(f"Extra for key 'foo': {foo_extra}")
+
+# ... your code that computes and persists the asset
+
+# Stream arbitrary metadata back to Dagster. This will be attached to the
+# associated `AssetMaterialization`
+context.report_asset_metadata("some_metric", get_metric(), metadata_type="text")
+
+# Stream data version back to Dagster. This will also be attached to the
+# associated `AssetMaterialization`.
+context.report_asset_data_version(get_data_version())
+```
+
+### (2) `ext_protocol` context manager
+
+Internally, `ExtDatabricks` is using the `ext_protocol` context manager to set
+up communications. If you'd prefer more control over how your databricks job is
+launched and polled, you can skip `ExtDatabricks` and use this lower level API
```

Comment:
If you'd prefer more control over how your databricks job or you have existing code to launch the job you do not want to change